packages/caravan-fees/src/cpfp.ts

```diff
@@ -0,0 +1,301 @@
+ *                 the combined fee rate doesn't meet the target (in strict mode).
+ *
+ * @example
+ * const cpfpTx = createCPFPTransaction({
+ *   originalTx: originalTxHex,
+ *   availableInputs: availableUTXOs,
+ *   spendableOutputIndex: 1,
+ *   changeAddress: 'bc1q...',
+ *   network: Network.MAINNET,
+ *   dustThreshold: '546',
+ *   scriptType: 'P2WSH',
+ *   targetFeeRate: '15',
+ *   absoluteFee: '1000',
+ *   requiredSigners: 2,
+ *   totalSigners: 3,
+ *   strict: true
+ * });
+ *
+ * @see https://bitcoinops.org/en/topics/cpfp/ Bitcoin Optech on Child Pays for Parent
+ * @see https://github.com/bitcoin/bitcoin/pull/7600 Bitcoin Core CPFP implementation
+ */
+export const createCPFPTransaction = (options: CPFPOptions): string => {
+  const {
+    strict = false,
+    originalTx,
+    network,
+    targetFeeRate,
+    absoluteFee,
+    availableInputs,
+    requiredSigners,
+    totalSigners,
+    scriptType,
+    dustThreshold,
+    spendableOutputIndex,
+    changeAddress,
+  } = options;
+
+  // CPFP Calculation Process:
+  // Step 1: Analyze the original transaction
+  const txAnalyzer = new TransactionAnalyzer({
+    txHex: originalTx,
+    network,
+    changeOutputIndex: spendableOutputIndex,
+    targetFeeRate, // We need this param ... very essential for the analyzer to make decisions cannot put it to 0
+    absoluteFee,
+    availableUtxos: availableInputs,
+    requiredSigners,
+    totalSigners,
+    addressType: scriptType,
+  });
+
+  const analysis = txAnalyzer.analyze();
+  // Step 2: Verify CPFP suitability
+  if (!analysis.canCPFP) {
+    throw new Error(
+      "CPFP is not possible for this transaction. Ensure the specified output is available for spending.",
+    );
+  }
+
+  if (analysis.recommendedStrategy !== FeeBumpStrategy.CPFP) {
+    console.warn(
+      "CPFP is not the recommended strategy for this transaction. Consider using the recommended strategy: " +
+        analysis.recommendedStrategy,
+    );
+  }
+
+  // Step 3: Create a new transaction template for the child transaction
+  const childTxTemplate = new BtcTransactionTemplate({
+    inputs: [],
+    outputs: [],
+    targetFeeRate: Number(txAnalyzer.cpfpFeeRate), // Use CPFP-specific fee rate from analyzer
+    dustThreshold,
+    network,
+    scriptType,
+    requiredSigners,
+    totalSigners,
+  });
+
+  // Step 4: Add the spendable output from the parent transaction as an input
+  const parentOutput = txAnalyzer.outputs[spendableOutputIndex];
+  if (!parentOutput) {
+    throw new Error(
+      `Spendable output at index ${spendableOutputIndex} not found in the parent transaction.`,
+    );
+  }
+
+  const parentOutputUTXO = availableInputs.find(
+    (utxo) => utxo.txid === analysis.txid && utxo.vout === spendableOutputIndex,
+  );
+
+  if (!parentOutputUTXO) {
+    throw new Error(
+      `UTXO for the spendable output (${analysis.txid}:${spendableOutputIndex}) not found in availableInputs.`,
+    );
+  }
```

Comment:
this feels like duplicative work no? If we have the parentOutput then we have the parentOutputUtxo. I'm not sure we need to require that it's in the available inputs since that's implied. 

I would simplify this to check
* is there a spendableOutputIndex passed to the function?
* is there an output in the parent transaction that matches? 
* If so, then set the parentOutputUtxo to that value (we should have the amount and txid already

Reply:
> If we have the parentOutput then we have the parentOutputUtxo

Not actually, we depend on the user code to provide us with all available inputs and so that's why we first find the output in the parent transaction and then check if that output (in form of UTXO) is provided to us in the available inputs 

also 
> is there a spendableOutputIndex passed to the function?
This check is handled by the txAnalyzer as :

```ts
  const txAnalyzer = new TransactionAnalyzer({
    txHex: originalTx,
    network,
    changeOutputIndex: spendableOutputIndex,
    ```
    
    here the analyzers validate options method does the job for us in the constructor

Reply:
> we depend on the user code to provide us with all available inputs 

I guess that's what I'm saying. I don't think we need to rely on the user code to provide the inputs that we already know about. In the case of cpfp, the input we know about is the spendable output from the parent tx. The user code should just be providing the ADDITIONAL inputs, but I think it's ok if the user doesn't provide the input that is coming from the parent tx if that makes sense.

Reply:
> I guess that's what I'm saying.
Ahh ... yes sorry,  now I understand your point ....

So actually the `txAnalyzer.outputs` contains outputs as `BtcTxOutputTemplate` type but to build our transaction we need a `UTXO` type which includes other information that we need for building the transaction as we'll be adding this output as input for our new child Tx ...

```ts
childTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(parentOutputUTXO));
```

If you remember we had a detailed discussion about this on discord where we concluded that we'll need extra fields like prevTxHex , etc and so it should be supplied by the user code itself ...

here is chunk of that discussion 

> So if we can assume that the user of the function has access to the full prevtx, I think it's reasonable to expect that you have that info when adding the input templates. I would put that as optional for creating the inputs, and then required for the rbf and cpfp functions so when you call them and build the templates with taht data, you have enough to build the psbt from a template and then throw an error if toPsbt is called without that info on the templates (maybe as part of a validate call).

Reply:
Though I too feel that we need to find another way to reduce dependence on user code to provide us with all the information and maybe use caravan-client package in future to access the info that we need ... thus simplifying the rbf/cpfp function calls too

Reply:
ahh .... understood ... thanks for the clarification 

So we are looking at something like this ...

```ts
 // Create a UTXO from the parent transaction's output
  const parentOutputUTXO2: UTXO = {
    txid: analysis.txid,
    vout: spendableOutputIndex,
    value: parentOutput.amountSats,
    witnessUtxo: {
      script: createOutputScript(parentOutput.address, network),
      value: parseInt(parentOutput.amountSats),
    },

    prevTxHex: originalTx,
  };
```

Reply:
Yep!

Reply:
changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/e3fe44158c5b04e78eed4fe72b2bdaf959579af0#diff-852fd83c241ed78594dc29389ec04d1425d784e665c706434a8de046f7daa85dR128)